src/components/GenericDropdown.tsx

```diff
@@ -15,38 +14,87 @@ const DropdownButton = styled.button`
 	border: 1px, solid, black;
 `;
 
+const DropdownMenu = styled.ul`
+	list-style: none;
+	background-color: ${({ theme }) => theme.colors.main[0]};
+	text-align: start;
+	z-index: 2;
+	position: absolute;
+	border: 1px solid rgba(0, 0, 0, 0.04);
+	box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14);
+	padding: 0;
+	margin: 0;
+	border-radius: 10px;
+`;
+
 const Container = styled.div`
 	display: flex;
 	flex-direction: column;
 	height: 100%;
 	width: 100%;
 `;
 
-function GenericDropdown({ buttonDisplay, dropdownItems }: Props) {
+export const ListLink = styled(Link)`
+	display: block;
+	height: 100%;
+	text-decoration: none;
+	color: ${({ theme }) => theme.colors.text[2]};
+	padding: 1em;
+	&:hover {
+		color: ${({ theme }) => theme.colors.primary[0]};
+	}
+`;
+
+export const ListButton = styled.div`
+	display: block;
+	height: 100%;
+	text-decoration: none;
+	color: ${({ theme }) => theme.colors.text[2]};
+	padding: 1em;
+	&:hover {
+		color: ${({ theme }) => theme.colors.primary[0]};
+	}
+`;
+
+function GenericDropdown({
+	buttonDisplay,
+	children,
+}: React.PropsWithChildren<Props>) {
 	const ref = useRef<HTMLDivElement>(null);
 	const buttonRef = useRef<HTMLButtonElement>(null);
-	const { isComponentVisible, setIsComponentVisible } = useComponentVisible(
+	const [componentVisible, setComponentVisible] = useComponentVisible(
 		ref,
 		false
 	);
 	const handleClick = () => {
-		setIsComponentVisible(!isComponentVisible);
+		setComponentVisible(!componentVisible);
 	};
 
+	const [top, setTop] = useState(-9999);
+
+	useEffect(() => {
+		if (buttonRef.current != null) {
+			const rect = buttonRef.current.getBoundingClientRect();
+			setTop(rect.bottom);
+		}
```

Comment:
So I agree about cleaning this up, and I think it was less confusing when there were more components as the context of the code provided a lot of insight. 

I think a custom hook is interesting, but I don't see it being particularly generic. Meaning no other component needs to be displayed floating below another component. That being said, what are your thoughts about defining a hook outside(above) the component? 

I think what I have written now is more clear, but probably some room for improvement. I think a custom hook though is kind of overkill for this situation...

Reply:
It's good to think in terms of reuse; it would be fine to pull the state and lifecycle logic into its own hook for the sake of encapsulation and as a separation of concerns. `GenericDropdown` shouldn't necessarily be concerned with the code that makes the dropdown work, ie. the hooks.

On one hand we have view logic, the code that handles displaying data with styling. On the other hand we have a form of business logic, the code that handles state and manipulation of the viewport using the `Element` class. These two things work hand-in-hand but they don't necessarily need to be defined in the same place.

If testing was a concern, separating the hook would also allow us to test the hook in isolation. As it stands, tests for `GenericDropdown` would have to implicitly test the lifecycle/state logic.

What I personally wouldn't do is define the hook in the same file; this is a bit of an anti-pattern, much like having two components built in one file.

The extra level of abstraction could be seen as overkill sure, but at first glance I personally don't know what this lifecycle hook is doing. At the very least it wouldn't hurt anyone and has the added benefit of possibly cleaning up the code and improving readability.